text/0031-expose-all-layer-metadata.md

```diff
@@ -0,0 +1,138 @@
+# Meta
+[meta]: #meta
+- Name: Include all layer metadata in final image
+- Start Date: 2020-03-12
+- CNB Pull Request: 
+- CNB Issue: 
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+Currently, the only layer metadata that is included in a final image
+is the layer metadata for that image's launch layers. It would benefit
+image debugging and build reproducibility to include _all_ layer
+metadata in the final image.
+
+# Motivation
+[motivation]: #motivation
+
+At the highest level, non-launch layer metadata can help us to know
+things about non-launch layers that participated in an image build,
+which things we might need to know in order to reproduce a
+build. Buildpacks users might want to know things about such
+non-launch layers when they discover a problem in an image, or a
+problem in a build; some such problems will depend on non-launch
+layers that participated in the build.
+
+Having implemented this feature, buildpack users and authors will have
+more information with which to debug and reproduce builds.
+
+Consider a build that can only be reproduced if the state of the cache
+is reproduced (the `@` notation means some git repo at some SHA):
+
+1. I build `some-app @ 000000`.
+2. I build `some-app @ 111111`.
+3. I notice a problem in the build for `some-app @ 111111`.
+4. I build `some-app @ 111111` without a cache, but cannot reproduce
+   the problem.
+5. I look at the layer metadata for the image associated with
+   `some-app @ 111111` and find that all of the layers used in its
+   weird build were generated using `some-app @ 000000`.
+6. I build `some-app @ 000000` without a cache (to "seed" the cache).
+7. I build `some-app @ 111111` with the cache from `some-app @ 000000`
+   and reproduce the issue.
+
+This example conflates the related issue of knowing what version of an
+application a particular layer arose from; however, the first step
+towards surfacing that information could be including all layer
+metadata in an image. Given that the layer metadata will be preserved,
+the buildpack author user or the platform could add the application's
+version (e.g. a git SHA) to the layer metadata.
```

Comment:
The lifecycle, as of Platform API 0.3, does currently provide a (poorly document) mechanism for platforms to set source metadata on the image (https://github.com/buildpacks/rfcs/blob/master/text/0013-app-layer-metadata-source.md, buildpacks/lifecycle#230)

Reply:
I'm sorry, but it's not entirely clear to me how you mean to relate
that lifecycle feature to this RFC. To me, that lifecycle feature
seems to answer a related question, "What version of the app did I use
as input when I made this image?" But in general, the layers used in
that image may have been made with different versions of the same app.

For example, consider an app foo-api and a git repo with the same
name, and say I've already built `registry.example.com/foo-api:000000`
with the the foo-api repo @ `000000`.
1. Now say I build `registry.example.com/foo-api:111111` with the the
   foo-api repo @ `111111`, and all of my cached layers from the
   previous build (for the image
   `registry.example.com/foo-api:000000`) are valid.
1. As a result of that build, I've got an
   `registry.example.com/foo-api:111111` image whose app source code
   comes from the sha `111111`, but whose layers were all generated
   using the app as of the sha `000000`.

In this case, I believe that `registry.example.com/foo-api:111111`
would have a label `Config.Labels."io.buildpacks.app.source"` that
identifies the app source code as coming from the sha `111111`, but
there's nothing to tell me that the launch layers (let alone
non-launch layers) were generated with the sha `000000`. So, that's
the issue that I'm trying to relate.

This example actually does come my team's use of buildpacks tools.

To try to relate this to your concept of 'source metadata', I might
say that there are actually many sources involved in image builds,
perhaps the most obvious two being the application source code and the
cached layers. In addition to having 'source metadata' at the image
level, to say what version of the application source code the image
contains, it's also valuable to have 'source metadata' at the layer
level to say what version of the application source code was on-disk
when a buildpack generated that layer.

Reply:
One reason I am resisting the idea of adding source metadata to the layer metadata, is that the source metadata isn't necessarily available to the buildpack at build time. One example: once `pack` supports `include` and `exclude` in `project.toml`, the `.git` directory is a likely candidate to be excluded.

I see that point you  are making, the source may influence what was created in a given layer. I think there is almost always a more direct way to express the relevant info. To take your `node_modules` example, the hash of the `package-lock.json` can be used to determine whether an existing cache layer containing `node_modules` is valid or if needs to be updated. The buildpack can already use the layer metadata to prevent a layer withoutdated models from being added to a new image. You could take the visibility a step further and add the actual contents of the resolved tree to the layer metadata (assuming this RFC get's accepted) or to the bill-of-materials (right now). If I am debugging an issue, I want to know what state at build time was relevant to the layer generation, without having to trace that back to the git sha of the last commit that changed the `package-lock.json`, especially since many other unrelated but important changes could have happened since then. If each of my cache-only layers came from a different sha, it wouldn't be easy to re-seed the cache using that information.

I really like the idea of adding all layer metadata to the image, but I think source metadata in particular, as an example, introduces distracting complications. Nothing would stop your buildpack from adding source metadata if you want it (and it is always available in your use cases) but I don't think it's the cleanest example for the RFC.

Reply:
My team's nodejs buildpack does use a content digest of
`package-lock.json` to validate its cache. That value has its place,
but it's no good for doing a reverse look-up; rather than require any
sort of reverse look up, I think that the layer should tell a user
exactly where it came from. For me, a checksum of `package-lock.json`
is not a more direct expression; perhaps I haven't fully under your
case.

I'm sure that I don't understand what you mean here:

>You could take the visibility a step further and add the actual
>contents of the resolved tree to the layer metadata (assuming this
>RFC get's accepted) or to the bill-of-materials (right now).

I do agree with you on this point:

>If I am debugging an issue, I want to know what state at build time
>was relevant to the layer generation, without having to trace that
>back to the git sha [...]

However, what if it was a bug in the buildpack that lead to some bad
cache? In that case, it would be useful to know the git SHA that was
used to generate the bad cache, and the version of the buildpack, to
reproduce the issue.

The use of git SHA that I'm trying to describe isn't about validating
the cache. It's about knowing how to reproduce the state of the cache
exactly in order to debug issues that arise when buildpacks, the
cache, and the application come together.

It's not ideal to have to checkout an old SHA and build it, but
sometimes it will be necessary to do so, particularly if the problem
exists in the buildpack code. I'm not recommending that anyone use
this as their go-to technique for debugging buildpack issues, but
sometimes there's nothing else that can be done. In general, I think
that there will be many debugging cases that would benefit from
knowing the version of the application in addition to the version of
the buildpack.

Layer source metadata is indeed distracting and complicating; if I
could add one last complication, it'd be that I don't want to have to
add source metadata for each of my layers - I just want to get that
for free, without adding boilerplate to each of my buildpacks.

That said, even now I have some doubts about making this reply,
because it's not clear to me that we're understanding one another.

Reply:
> the source metadata isn't necessarily available to the buildpack at build time. One example: once pack supports include and exclude in project.toml, the .git directory is a likely candidate to be excluded.

To me, this is definitely a problem with the use case. But I don't have a strong opinion about what example is used in the RFC, even though it seems like a more simple example would suffice.